File: DFS, BFS/7562.py
#나이트의 이동
from collections import deque
def bfs(x, y):
    queue = deque()
    queue.append((x, y))
    while queue:
        x, y = queue.popleft()
        for i in range(8):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0<=nx<l and 0<=ny<l and graph[nx][ny] == 0:
                graph[nx][ny] = graph[x][y] + 1
                queue.append((nx, ny))
    return graph[go_x][go_y] - 1

dx = [-2, -1, 1, 2, 2, 1, -1, -2]
dy = [1, 2, 2, 1, -1, -2, -2, -1]
t = int(input())
for _ in range(t):
    graph = []
    l = int(input())
    cur_x, cur_y = map(int, input().split())
    go_x, go_y = map(int, input().split())
    graph = [[0 for _ in range(l)] for _ in range(l)]
    graph[cur_x][cur_y] = 1
    print(bfs(cur_x, cur_y))

File: DFS, BFS/7562.js
// 나이트의 이동
let input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

let t = Number(input[0]);
let idx = 1;
let dx = [-2, -1, 1, 2, 2, 1, -1, -2];
let dy = [1, 2, 2, 1, -1, -2, -2, -1];

for (let v = 0; v < t; v++) {
  let n = Number(input[idx++]);
  let [sx, sy] = input[idx++].split(" ").map((v) => Number(v));
  let [gx, gy] = input[idx++].split(" ").map((v) => Number(v));
  let graph = Array.from(Array(n), () => Array(n).fill(0));
  bfs(graph, n, sx, sy, gx, gy);
  console.log(graph[gx][gy]);
}

function bfs(graph, n, sx, sy, gx, gy) {
  if (sx === gx && sy === gy) return;
  let q = [];
  q.push([sx, sy]);
  while (q.length) {
    let [x, y] = q.shift();
    for (let i = 0; i < 8; i++) {
      let nx = x + dx[i];
      let ny = y + dy[i];
      if (0 <= nx && nx < n && 0 <= ny && ny < n && graph[nx][ny] === 0) {
        graph[nx][ny] = graph[x][y] + 1;
        q.push([nx, ny]);
      }
    }
  }
}
